timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java

```diff
@@ -0,0 +1,190 @@
+        try {
+          MappingIterator<Forex> object = objectReader.readValues(input);
+          if (object.hasNext()) {
+            Forex forex = null;
+            try {
+              forex = object.next();
+
+              if (convertCSVForex.tickers.contains(forex.ticker)) {
+
+                TSKey key = TSKey.newBuilder().setMajorKey(forex.ticker).build();
+
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSSSSS");
+
+                LocalDateTime ldt = LocalDateTime.parse(forex.timestamp, formatter);
+                // Setting UTC using Reykjavik time zone
+                ZonedDateTime zdt = ldt.atZone(ZoneId.of("Atlantic/Reykjavik"));
+
+                long millis = zdt.toInstant().toEpochMilli();
+
+                // OUTPUT Ask Price
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("ASK"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.ask))
+                        .build(), Instant.ofEpochMilli(millis));
+
+                // OUTPUT Bid Price
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("BID"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.bid))
+                        .build(), Instant.ofEpochMilli(millis));
+
+                // OUTPUT Ask volume
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("ASK_VOLUME"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.ask_volume))
+                        .build(), Instant.ofEpochMilli(millis));
+
+                // OUTPUT Bid volume
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("BID_VOLUME"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.bid_volume))
+                        .build(), Instant.ofEpochMilli(millis));
+              }
+            } catch (Exception ex) {
+              System.out.println(String.format("Unable to parse record: %s", ex));
+            }
+          }
+
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+  }
+
+  /**
+   * ticker|timestamp|ask|bid|ask_volume|bid_volume Intermediary POJO used for conversion of CSV
+   * file.
+   */
+  @DefaultSchema(JavaFieldSchema.class)
```

Comment:
Do we expect all these to always be not null?

Reply:
I would say yes, but the logic includes now checks if some of the values failed to be converted, sending these elements to the deadletter queue.